source/MaterialXRenderGlsl/TextureBaker.cpp

```diff
@@ -310,7 +233,220 @@ void TextureBaker::optimizeBakedTextures(NodePtr shader)
     }
 }
 
-DocumentPtr TextureBaker::bakeMaterial(NodePtr shader, const StringVec& udimSet)
+DocumentPtr TextureBaker::bakeMaterialToDoc(DocumentPtr doc, const FileSearchPath& searchPath, const string& materialPath, 
+                                              const StringVec udimSet, string& documentName)
+{
+    // Set up generator context for material
+    GenContext genContext(_generator);
+    genContext.getOptions().targetColorSpaceOverride = LIN_REC709;
+    genContext.getOptions().fileTextureVerticalFlip = true;
+    genContext.getOptions().targetDistanceUnit = _distanceUnit;
+
+    DefaultColorManagementSystemPtr cms = DefaultColorManagementSystem::create(genContext.getShaderGenerator().getTarget());
+    cms->loadLibrary(doc);
+    for (const FilePath& path : searchPath)
+    {
+        genContext.registerSourceCodeSearchPath(path / "libraries");
+    }
+    genContext.getShaderGenerator().setColorManagementSystem(cms);
+
+    // Compute the material tag set.
+    StringVec materialTags = udimSet;
+    if (materialTags.empty())
+    {
+        materialTags.push_back(EMPTY_STRING);
+    }
+
+    if (_outputStream)
+    {
+        *_outputStream << std::endl << "Working on material: " << materialPath << std::endl;
+    }
+
+    NodePtr shaderNode;
+    // Iterate over material tags.
+    for (const string& tag : materialTags)
+    {
+        StringResolverPtr resolver = StringResolver::create();
+        resolver->setUdimString(tag);
+        std::string resolvedRenderablePath = materialPath;
+
+        // Configures texture baking for materials that explicitly contain the <UDIM> token and will be baked as a group. 
+        if (!tag.empty() && materialPath.find(UDIM_TOKEN) != string::npos)
+        {
+            // Determine material group.
+            string udimPrefix = (_texTemplateOverrides.count("$UDIMPREFIX")) ? _texTemplateOverrides["$UDIMPREFIX"] : UDIM_PREFIX;
+            resolver->setUdimString(UDIM_TOKEN);
+            resolver->setFilenameSubstitution(udimPrefix + UDIM_TOKEN, "");
+            _materialGroupName = resolver->resolve(FilePath(materialPath).getBaseName(), FILENAME_TYPE_STRING);
+
+            // Determine explicit material renderable path with the current udim tag.
+            resolver->setUdimString(tag);
+            resolver->setFilenameSubstitution(udimPrefix + UDIM_TOKEN, udimPrefix + UDIM_TOKEN);
+            resolvedRenderablePath = resolver->resolve(materialPath, FILENAME_TYPE_STRING);
+        }
+        ElementPtr elem = doc->getDescendant(resolvedRenderablePath);
+        if (!elem || !elem->isA<Node>())
+        {
+            continue;
+        }
+        NodePtr materialNode = elem->asA<Node>();
+
+        vector<NodePtr> shaderNodes = getShaderNodes(materialNode);
```

Comment:
Could this computation of `shaderNode` be moved outside of the material tag loop?  It seems slightly odd for the variable to be declared outside the loop but defined internally, leaving the last definition of the variable to be used by subsequent logic.